Added TextCNN wrapper TorchModel

Shiva-sankaran opened the pull request:
## Description

Following up on #3828. Second PR for #3828 , adds TextCNN wrapper TorchModel.

<!-- Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change. -->


## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

rbharath reviewed (commented):
Can you please review @shreyasvinaya

rbharath commented on deepchem/models/torch_models/text_cnn.py:
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
         return outputs
+
+
+class TextCNNModel(TorchModel):
```
Let's add this to the docs/

Shiva-sankaran replied:
@shreyasvinaya I have a small doubt regarding updating docs. There is already an entry in the documentation for the TextCNNModel(https://deepchem.readthedocs.io/en/latest/api_reference/models.html#textcnnmodel), which currently links back to the TF implementation, should I replace it with this Torch version.

Shiva-sankaran replied:
Also do I replace Keras with PyTorch in this table(https://deepchem.readthedocs.io/en/latest/api_reference/models.html#id134) as well? I don't see any entries of Keras/PyTorch

shreyasvinaya replied:
add this to the docs under `torch_models`

Shiva-sankaran replied:
Added to docs.

shreyasvinaya commented on deepchem/models/torch_models/text_cnn.py:
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
+            mode: str = 'fit',
+            deterministic: bool = True,
+            pad_batches: bool = True) -> Iterable[Tuple[List, List, List]]:
+        """Transfer smiles strings to fixed length integer vectors"""
```
Documentation on parameters to be added

Shiva-sankaran replied:
Added parameter descriptions.

shreyasvinaya commented on deepchem/models/torch_models/text_cnn.py:
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
+    def build_char_dict(dataset: Dataset,
+                        default_dict: Dict[str, int] = default_dict):
+        """ Collect all unique characters(in smiles) from the dataset.
+       This method should be called before defining the model to build appropriate char_dict
```
Documentation on parameters to be added

Shiva-sankaran replied:
Added parameter descriptions.

shreyasvinaya commented on deepchem/models/torch_models/text_cnn.py:
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
+
+    def smiles_to_seq(self, smiles: str):
+        """
+        Tokenize characters in smiles to integers
```
Documentation on parameters to be added

Shiva-sankaran replied:
Added parameter descriptions.

shreyasvinaya commented on deepchem/models/torch_models/text_cnn.py:
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
+        return np.array(seq, dtype=np.int32)
+
+    @staticmethod
+    def convert_bytes_to_char(s):
```
Documentation on parameters to be added
type annotations missing

Shiva-sankaran replied:
Added parameter descriptions and type annotations.

shreyasvinaya commented on deepchem/models/torch_models/text_cnn.py:
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
+        s = ''.join(chr(b) for b in s)
+        return s
+
+    def smiles_to_seq_batch(self, ids_b):
+        """
+        Converts SMILES strings to np.array sequence.
+        """
```
Documentation on parameters to be added
type annotations missing

Shiva-sankaran replied:
Added parameter descriptions and type annotations.

rbharath commented:
@Shiva-sankaran Before requesting a new review, make sure to respond to all remaining comments on the PR (can be thumbs up, or a sentence response even) so we know you have addressed all comments

Shiva-sankaran commented:
> @Shiva-sankaran Before requesting a new review, make sure to respond to all remaining comments on the PR (can be thumbs up, or a sentence response even) so we know you have addressed all comments

Understood.

rbharath commented on docs/source/api_reference/models.rst:
```diff
@@ -299,7 +299,7 @@ WGAN
 TextCNNModel
 ------------
 
-.. autoclass:: deepchem.models.TextCNNModel
```
Don't remove the old TF model; add this as a new model

Shiva-sankaran replied:
Reverted replacing the old TF model. Added the docs for the torch model under the Torch Models section.

rbharath reviewed (commented):
One comment to fix on the docs and please rebase the PR

rbharath commented:
@Shiva-sankaran Please don't merge; rebase instead. Merging messes up large PRs. Can you please rebase?

Shiva-sankaran commented:
> @Shiva-sankaran Please don't merge; rebase instead. Merging messes up large PRs. Can you please rebase?

Yes, learnt it the hard way. Fixed it.

rbharath reviewed (commented):
@shreyasvinaya if this model looks good to you, can you LGTM it? I will go ahead and merge it in if it looks good to you

shreyasvinaya commented on docs/source/api_reference/molecular_models.csv:
```diff
@@ -11,7 +11,7 @@ MEGNetModel,`ref <https://arxiv.org/abs/1812.05055>`_,Classifier/ Regressor,,PyT
 PagtnModel,`ref <https://arxiv.org/abs/1905.12712>`_,Classifier/ Regressor,PagtnMolGraphFeaturizer MolGraphConvFeaturizer,DGL/PyTorch,
 Smiles2Vec,`ref <https://arxiv.org/abs/1712.02034>`_,Classifier/ Regressor,SmilesToSeq,Keras,
-TextCNNModel,`ref <https://arxiv.org/abs/1705.10843>`_,Classifier/ Regressor,,Keras,
+TextCNNModel,`ref <https://arxiv.org/abs/1705.10843>`_,Classifier/ Regressor,,PyTorch,
```
Can you convert this to `Keras/PyTorch` as both implementations exist in the repo

Shiva-sankaran replied:
Done.

shreyasvinaya commented:
LGTM, ready to merge once the small docfix is done

shreyasvinaya reviewed (approved):
LGTM once CI passes
cc @rbharath

rbharath commented:
@shreyasvinaya I'm seeing errors in docs-build python 3.11. Are these linked to this PR?

shreyasvinaya commented:
@Shiva-sankaran can you please rebase the PR, i think the new commits should fix the doctest issues

Shiva-sankaran commented:
@shreyasvinaya Rebased, there is still 2 doctest failures.

shreyasvinaya commented:
@rbharath LGTM, good to merge

rbharath reviewed (approved):
LGTM
